### PR TITLE
Adding the goal conditioning sensors with the new observation specs

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Adding the goal conditioning sensors with the new observation specs

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Sensors/ISensor.cs [hunk 1]
```diff
@@ -59,16 +59,6 @@ public enum ObservationType
         /// Collected observations contain goal information.
         /// </summary>
         Goal = 1,
-
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/ISensor.cs:62]: We were going to rename "Goal" to "GoalSignal" too, right?

File: com.unity.ml-agents/Runtime/Sensors/VectorSensor.cs [hunk 2]
```diff
@@ -21,16 +21,20 @@ public class VectorSensor : ISensor, IBuiltInSensor
...
-        public VectorSensor(int observationSize, string name = null)
+        public VectorSensor(int observationSize, string name = null, ObservationType observationType = ObservationType.Default)
         {
-            if (name == null)
+            if (name == null || name == "")
             {
                 name = $"VectorSensor_size{observationSize}";
+                if (observationType != ObservationType.Default)
+                {
+                    name += "_goal";
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/VectorSensor.cs:31]:
```suggestion
                    name += $"_{observationType.ToString()}";
```
(double check that compiles)

File: com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        private string m_SensorName = "VectorSensor";
+
+        public int ObservationSize
+        {
+            get { return m_observationSize; }
+            set { m_observationSize = value; }
+        }
+
+        [HideInInspector, SerializeField]
+        int m_observationSize;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs:28]: m_ObservationSize

File: com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        [HideInInspector, SerializeField]
+        ObservationType m_ObservationType;
+
+        VectorSensor m_sensor;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs:33]: m_Sensor

File: com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs (hunk 3 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs:28]: (and don't forget to change the string in VectorSensorComponentEditor)

**Review by chriselion — APPROVED** [review]
Looks good pending the changes.

File: com.unity.ml-agents/Runtime/Sensors/ISensor.cs (hunk 1 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/ISensor.cs:62]: I forgot this conversation. :( I can change it!

File: com.unity.ml-agents/Runtime/Sensors/VectorSensor.cs [hunk 5]
```diff
@@ -21,16 +21,20 @@ public class VectorSensor : ISensor, IBuiltInSensor
         /// </summary>
         /// <param name="observationSize">Number of vector observations.</param>
         /// <param name="name">Name of the sensor.</param>
-        public VectorSensor(int observationSize, string name = null)
+        public VectorSensor(int observationSize, string name = null, ObservationType observationType = ObservationType.Default)
         {
-            if (name == null)
+            if (name == null || name == "")
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/VectorSensor.cs:26]:
```suggestion
            if (string.IsNullOrEmpty(name))
```

File: com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
+using UnityEngine;
+using UnityEngine.Serialization;
+
+namespace Unity.MLAgents.Sensors
+{
+    [AddComponentMenu("ML Agents/Vector Sensor", (int)MenuGroup.Sensors)]
+    public class VectorSensorComponent : SensorComponent
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs:7]: Please add docstrings here and on public properties.

Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensors/VectorSensorComponent.cs:10]: Ok, added the docstring here and on the properties.

**Pull request merged**